### DISCUSSION TITLE
Shallow Routing on a dynamic page/route

### DISCUSSION THREAD

**Discussion by juliuslissau** [vercel/next.js#40256] [Help] [answered]
From a normal page, I can do `router.push("/this-page", "/this-page/" + id, {shallow: true})` without refreshing the page. 

I'd like to be able to do that, from a dynamic route (e.g. [...city].js) as well.

This unfortunately creates a full refresh instead: `router.push("/this-page/" + city, "/this-page/house/" + id, {shallow: true})`

Does anyone know how to fix/avoid this issue?

I made a demo of my issue on Stackblitz here: https://stackblitz.com/edit/nextjs-ezyusx?file=pages/index.js

**Comment by icyJoseph** [comment]
I am not sure on what are you trying to do here:

```js
  const changeRoute = (id) => {
    router.push(
      `/dynamic-post/${router?.query?.category}`,
      `/dynamic-posts/post/${id}`,
      {
        shallow: true,
      }
    );
  };
```

Router push regularly just works as a `go-to` function, meaning that you call it `router.push('/go-here')`, the reason why there's a second argument is **not** for a `from` here, go `there` situation, but rather, to match a file system path, to a browser url.

The [OLD DOCS](https://nextjs.org/docs/tag/v9.5.2/api-reference/next/link#dynamic-routes) explain this very well. In general you can pass `null`, or `undefined` to the second argument if you don't need it, or if you want to pass options, but don't want to mess with the second argument.

Under some circumstances, the `href` + `as` combo is still necessary, but it usually for highly dynamic stuff. 

There's also a typo, in your application `/dynamic-post` doesn't exist as a route. I do believe this is why your page refreshes, but then somehow the `as` argument, shown in the browser, makes it recover itself...

This should do the trick:

```jsx
  const changeRoute = (id) => { 
    // null or undefined works alright
    router.push(`/dynamic-posts/post/${id}`, null, {
      shallow: true,
    });
  };
```

Now the question is why use `shallow: true`? Are you thinking that shallow true prevents page refresh? That's not at all what it does.

When you use `Link` from `next/link` or `push/replace` from `next/router` you are doing Single Page Application (SPA) transitions, which keep the page loaded, and use the Next.js router to determine what should be rendered to the screen.

The shallow option is to prevent calling `getServerSideProps` and friends, https://nextjs.org/docs/routing/shallow-routing:

> Shallow routing allows you to change the URL without running data fetching methods again, that includes [getServerSideProps](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props), [getStaticProps](https://nextjs.org/docs/basic-features/data-fetching/get-static-props), and [getInitialProps](https://nextjs.org/docs/api-reference/data-fetching/get-initial-props).

Further down on Shallow Routing there's a caveat about changing the page, and how that does trigger a page reload, even though, we didn't want it to. Notice that is kind of similar to your issue, but in your case even if you fix the typo, there's still no `category.js` on the file system so it'll just hit a 404.

**Reply by juliuslissau** [reply]
Thanks a lot for the answer!
I fixed the typo, it wasn't the reason for refresh.

I guess I'll be a bit more specific in what trying to achieve.
I'm building a webportal for houses like Zillow.

I need the following types of URLs:
1) website.com/houses (shows a list of all houses)
2) website.com/houses/new-york (shows all houses in a city)
3) website.com/houses/123 or website.com/houses/item/123 (a specific house)

I need all pages to be statically generated with getStaticPaths and getStaticProps for SEO. Server side rendering seems quite slow. 

Page 1 and 2 will be the same file preferably. So the folder would be "/houses/[[...city]].js". 
And then the getStaticPaths of that file would also include the id's of the individual houses, or there would be another "houses/items/[id].js" file.

So here's the user flow:
- When the user clicks on a house listed on page 1 above, then the house details opens in a modal (not a link to page 3).
- At the same time, I also need to be able to change the URL to page 3, without causing any redirect or refresh. 
(if the user then closes the modal, the URL will be changed back to either page 1 or 2 depending or where the user was)
- The URL just need to cosmetically change, so it's possible for the user to copy the URL to the specific house.

---

The problem is, that if I didn't need the city pages (page 2 from above), and just had website.com/houses and then website.com/houses/123 then I can just do router.push("/houses", "/houses/" + id, {shallow: true}) and then the URL cosmetically changes but no refresh or redirect.

But when I introduce the need for page 2 as described above (for SEO reasons), and change houses.js to a dynamic route like in the stackblitz example, then I can't just change the URL cosmetically.

**Comment by juliuslissau** [comment, marked as answer]
Found a solution, I can just use window.history.pushState() instead of using the router
